### ISSUE TITLE
New readability rule to format ternary expression

### ISSUE THREAD

**Issue by Przemyslaw-W** [DotNetAnalyzers/StyleCopAnalyzers#651]
Hi,

I'd like to propose new readability rule for ternary expressions. Valid options should be:

a) whole expression in one line:
```c#
    int x = booleanExpression ? firstIntValueExpression : secondIntValueExpression;
```
b) each subexpression on its own line, with "?" and ":" as starting characters:
```c#
    int x = booleanExpression
        ? firstIntValueExpression
        : secondIntValueExpression;
```
Rationale for placement expressions in separate lines: 
* consistency with parameters and statement rules - parameters should be in one line or each in separate lines, statements should be each in its own line. 
* each expression in its own line allows to easy grasp where it starts and ends.

Rationale for `?` and `:` placement: 
* when these operators are placed at the beginning of line, one can easy see that rest of the line is part of ternary expression. 
* when these operators are placed at the end of line and expressions are long, `?` and `:` tokens can go out of view. This makes harder to see that ternary expression is used.

Here is corresponding original SC issue, together with follow up discussion: https://stylecop.codeplex.com/discussions/253622

**Comment by AArnott** [comment]
:+1: I already follow this rule and ask my team to follow it. I very much agree it helps with readability. And if folks can't fit each *sub*-expression on a single line, maybe they shouldn't be using the ternary operators.

**Comment by sharwell** [comment]
:+1: I agree on this one too.

**Comment by sharwell** [comment]
Moved this one forward.

**Comment by vweijsters** [comment]
:+1: for me as well

@sharwell I think it would a good idea if you would assign a identifier to an accepted proposal (preferably in the title of the issue).

**Comment by otac0n** [comment]
var category = p == "Cat" ? "Animals" :
                   p == "Bat" ? "Equipment" :
                   p == "Fat" ? "Nutrition" : "None";

Should be allowed.

**Comment by sharwell** [comment]
It looks like you are trying to treat the ternary operator like a match operator from functional programming. A better form for the expression, and one that is compatible with this rule, is using either a `switch` statement or a sequence of `if`/`else` statements.

**Comment by otac0n** [comment]
So you are saying that this:

    var x = p == "Cat" ? "Animals" :
            p == "Bat" ? "Equipment" :
            p == "Fat" ? "Nutrition" : "None";

MUST be written as follows, in order to be compliant?:

    string x;
    if (p == "Cat")
    {
        x = "Animals";
    }
    else if (p == "Bat")
    {
        x = "Equipment";
    }
    else if (p == "Fat")
    {
        x = "Nutrition";
    }
    else
    {
        x = "None";
    }

Are you serious?  Do you not see how much harder that is to read and write? StyleCop is supposed to improve readability, not make it worse.  Consistency is a laudable goal, but never at the price of readability.

**Comment by AArnott** [comment]
I agree with @otac0n for the multiple expression case. Perhaps we can have the originally proposed format encouraged for single expressions and @otac0n's format for multi-expressions.
Unless parentheses are involved, in which case, I suggest stylecop just let it be however the developer wrote it.

**Comment by otac0n** [comment]
:+1:

**Comment by otac0n** [comment]
I should elaborate to say that what @AArnott described is exactly how I've been formatting my code for a couple of years.  I think it turns out very readable. (e.g. [single conditional](https://github.com/otac0n/GitReview/blob/master/GitReview/ActionResults/AdvertiseRefsResult.cs#L65 ) vs [multiple conditionals](https://github.com/otac0n/GitReview/blob/master/GitReview/ActionResults/ProtocolUtils.cs#L348))

**Comment by sharwell** [comment]
>  I think it turns out very readable.

I find the second example is good motivation for this rule without the exclusion. The following is more lines of code, but readability is *dramatically* improved.

```csharp
if (this.sourceIdentifier == null)
{
  return "create " + this.canonicalName + " => " + this.targetIdentifier;
}
else if (this.targetIdentifier == null)
{
  return "delete " + this.canonicalName + " (was " + this.sourceIdentifier + ")";
}
else
{
  return "update " + this.canonicalName + " => " + this.targetIdentifier + " (was " + this.sourceIdentifier + ")";
}
```

**Comment by otac0n** [comment]
Yeah, if I had been aligning horizontally, readability could have been improved:

                    return this.sourceIdentifier == null ? "create " + this.canonicalName + " => " + this.targetIdentifier :
                           this.targetIdentifier == null ? "delete " + this.canonicalName + " (was " + this.sourceIdentifier + ")" :
                                                           "update " + this.canonicalName + " => " + this.targetIdentifier + " (was " + this.sourceIdentifier + ")";

I do take your point, though.

Is there a simple rule we can use here that won't reduce the readability of existing code?  Perhaps, just keep the original proposal for single conditionals and ignore the rest?

**Comment by sharwell** [comment]
> Is there a simple rule we can use here that won't reduce the readability of existing code?

So far I haven't seen cases that would be negatively impacted by the original rule proposal, but I have seen cases that would be positively impacted. Once the rule is in the wild (in a beta) we'll have a much better idea of whether or not exclusions are actually needed.

**Comment by AArnott** [comment]
Just a thought: multiple check cases can often be rewritten as switch statements (not always, of course, since switch expressions must be one of very few primitive types).

**Comment by GregReddick** [comment]
It should be noted that they are talking about adding a match operator to C# version 7 as part of the standard C# syntax. https://github.com/dotnet/roslyn/issues/206 and https://github.com/dotnet/roslyn/blob/future/docs/features/patterns.md

**Comment by julealgon** [comment]
Folks, was this abandoned? It sounded like a fair proposal, but for some odd reason discussions abruptly stopped 3 years ago?

**Comment by SpaceKatt** [comment]
Any update on this issue? I want to enforce this rule.

**Comment by sharwell** [comment]
> So you are saying that this:
> 
> ```
> var x = p == "Cat" ? "Animals" :
>         p == "Bat" ? "Equipment" :
>         p == "Fat" ? "Nutrition" : "None";
> ```

Now that switch expressions are out, we can do this:

```cs
var x = p switch
{
  "Cat" => "Animals",
  "Bat" => "Equipment",
  "Fat" => "Nutrition",
  _ => "None",
};
```

**Comment by sharwell** [comment]
Marked up for grabs, again using the original proposal.

**Comment by AArnott** [comment]
I would rather see the original proposal modified to accommodate @otac0n's proposal, which for multiple ternary expressions allows this formatting:

```cs
var x = p == "Cat" ? "Animals" :
        p == "Bat" ? "Equipment" :
        p == "Fat" ? "Nutrition" : "None";
```

If the diagnostic is to recommend `switch` expressions, we should only recommend that when doing so would be valid C#. I believe the left side of the `=>` has to be a constant or pattern matching expression, which is a subset of what you can put in any expression that could appear in a ternary condition.

**Comment by julealgon** [comment]
> I believe the left side of the `=>` has to be a constant or pattern matching expression, which is a subset of what you can put in any expression that could appear in a ternary condition.

You would be incorrect there I think, in the sense that a pattern matching expression paired with a `when` allows you to perform arbitrary checks (same as you could with a ternary operator).

**Comment by AArnott** [comment]
Yes, I suppose you're right. But an automated code fix would have to know how to split an expression into the pattern matching part and the `when` part. Unless you just set all patterns to `true` and put the whole expression under `when`, but that doesn't feel like a good use of `switch` expressions, IMO.

**Comment by sharwell** [comment]
@AArnott I lead towards starting with the simple form of the rule and then evaluate it on real-world solutions to see if it needs alteration
